### PR TITLE
Add monitoring controller

### LINKED ISSUE #2124: Extract monitoring relevant components into dedicated controller
**User Story**
To reduce the scope of the cluster controller, monitoring related actions should be moved into a dedicated controller.

Deploying and modifying a Prometheus installation as well as kube-state-metrics are the only monitoring part we have for user clusters.
They do belong into the monitoring space - They have nothing to do with the lifecycle of a kubernetes cluster. We could remove them and all user clusters work as before.
The only purpose of the components(Prometheus, kube-state-metrics) is to utilize them in our monitoring.
They have specific configurations which can even be mofified by a kubermatic operator (Custom scrape configs & rules)

It would enable the seed team to manage all monitoring relevant logic - for user clusters - into a own controller - without interfering the with cluster lifecycle team.

From a failure domain perspective: As the prometheus configuration can be modified by a operator, it could potentially break the controller - killing the managemend of the user cluster (In case we have everything in one controller). Splitting each domain into a own controller will ensure that we provide the best service to clusters.

Going further down the road:
The current plan for the OpenShift integration is to have a dedicated controller which is responsible to manage the the OpenShift components for a user cluster.

There's still the plan to open source our cluster controller - but without all the extra bits we added. It should be a simple controller which only manages the control plane of Kubernetes.

Actions to be covered by this controller
- Deploy the Prometheus into the cluster namespace
- Modify Prometheus config based on given flags (Scrape config / alerting rules)
- Deploy kube-state-metrics into the cluster-namespace

**Acceptance criteria**
- [ ] A new controller has been added to the kubermatic controller-manager
- [ ] The cluster controller does not handle any monitoring relevant components
- [ ] Prometheus & kube-state-metrics get deployed via the new controller
- [ ] The new controller waits for the control plane to be running until it deploys the monitoring components (Check `cluster.status.health`)

### PR DESCRIPTION
**Pull request by thetechnick**
**What this PR does / why we need it**:
See #2124 
**tl;dr**: Move monitoring stuff into its own controller. The cluster controller should only need to handle cluster live-cycle relevant components.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2124 

**Special notes for your reviewer**:
I duplicated quite a lot of code, but I did not want to waste time on abstractions, before discussing it with someone in person.
I did not yet test if this is actually working, apart from a very basic unittest, because I did not get my hands on a test system yet. 

**Documentation**:

**Release note**:

*internal refactoring*
```release-note
NONE
```

### CODE REVIEW

File: api/cmd/kubermatic-controller-manager/controllers.go [hunk 1]
```diff
@@ -158,6 +160,46 @@ func createBackupController(ctrlCtx *controllerContext) (runner, error) {
...
+		ctrlCtx.runOptions.dc,
+		dcs,
+		ctrlCtx.runOptions.overwriteRegistry,
+		ctrlCtx.runOptions.nodePortRange,
+		ctrlCtx.runOptions.nodeAccessNetwork,
+		ctrlCtx.runOptions.etcdDiskSize,
+		ctrlCtx.runOptions.inClusterPrometheusRulesFile,
+		ctrlCtx.runOptions.inClusterPrometheusDisableDefaultRules,
+		ctrlCtx.runOptions.inClusterPrometheusDisableDefaultScrapingConfigs,
+		ctrlCtx.runOptions.inClusterPrometheusScrapingConfigsFile,
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-controller-manager/controllers.go:187]: Remove those arguments from the cluster controller: https://github.com/kubermatic/kubermatic/blob/monitoring-controller/api/pkg/controller/cluster/cluster_controller.go#L68

File: api/cmd/kubermatic-controller-manager/controllers.go [hunk 2]
```diff
@@ -158,6 +160,46 @@ func createBackupController(ctrlCtx *controllerContext) (runner, error) {
...
+	return monitoring.New(
+		ctrlCtx.kubeClient,
+		client.New(ctrlCtx.kubeInformerFactory.Core().V1().Secrets().Lister()),
+
+		ctrlCtx.runOptions.dc,
+		dcs,
+		ctrlCtx.runOptions.overwriteRegistry,
+		ctrlCtx.runOptions.nodePortRange,
+		ctrlCtx.runOptions.nodeAccessNetwork,
+		ctrlCtx.runOptions.etcdDiskSize,
```
Comment (mrIncompetent) [inline api/cmd/kubermatic-controller-manager/controllers.go:183]: Whats the disk size for?

File: api/pkg/controller/monitoring/monitoring_controller.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+	clusterRoleBindingInformer rbacv1informer.ClusterRoleBindingInformer,
+	deploymentInformer appsv1informer.DeploymentInformer,
+	secretInformer corev1informers.SecretInformer,
+) (*Controller, error) {
+	c := &Controller{
+		kubeClient: kubeClient,
+
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 5*time.Minute),
+			"monitoring",
```
Comment (mrIncompetent) [inline api/pkg/controller/monitoring/monitoring_controller.go]: `monitoring_cluster`

The name should be the name of the controller + the resource, which gets put into the queue.
Exception here is the cluster controller with just `cluster`.

File: api/cmd/kubermatic-controller-manager/controllers.go (hunk 1 above)
Comment (thetechnick) [reply api/cmd/kubermatic-controller-manager/controllers.go:187]: The prometheus arguments and the `ctrlCtx.runOptions.etcdDiskSize` are used here when instantiating a new TemplateData container:
https://github.com/kubermatic/kubermatic/blob/monitoring-controller/api/pkg/controller/cluster/resources.go#L88-L104
https://github.com/kubermatic/kubermatic/blob/monitoring-controller/api/pkg/controller/monitoring/resources.go#L18-L34

I am also not really happy with this solution, but I also find that `TemplateData` should always have all parameters, whether they are used or not (to make runtime errors less likely). 
Separating these TemplateData objects would make this PR bigger and increases the complexity to a point I am even less happy with.

File: api/cmd/kubermatic-controller-manager/controllers.go (hunk 2 above)
Comment (thetechnick) [reply api/cmd/kubermatic-controller-manager/controllers.go:183]: see answer above

**tl;dr** `NewTemplateData` needs the `etcdDiskSize`
https://github.com/kubermatic/kubermatic/blob/monitoring-controller/api/pkg/controller/monitoring/resources.go#L18-L34

File: api/pkg/controller/cluster/resources.go [hunk 4]
```diff
@@ -203,9 +196,7 @@ func (cc *Controller) ensureRoles(c *kubermaticv1.Cluster) error {
 }
 
 func (cc *Controller) ensureRoleBindings(c *kubermaticv1.Cluster) error {
-	creators := []resources.RoleBindingCreator{
-		prometheus.RoleBinding,
-	}
+	creators := []resources.RoleBindingCreator{}
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/resources.go]: I would prefer if we remove empty ensure functions

File: api/pkg/controller/monitoring/monitoring_controller.go [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	c.queue.Forget(key)
+	// Report to an external entity that, even after several retries, we could not successfully process this key
+	runtime.HandleError(err)
+	glog.V(0).Infof("Dropping %q out of the backup controller queue after exceeding max retries: %v", key, err)
```
Comment (mrIncompetent) [inline api/pkg/controller/monitoring/monitoring_controller.go]: `backup controller` -> `monitoring controller`

**Comment by mrIncompetent** [comment]
looks good from my side but @alvaroaleman should have a look as well. 
But reminds me of some cleanups we should do in our controllers (Not this PR):
- Fix adding back to the queue in case of error (Remove dropping)
- Eliminate global `TemplateData`. Maybe: https://github.com/kubermatic/kubermatic/compare/master...alternative-cm-creator

Really good PR :)

**Comment by mrIncompetent** [comment]
A potential solution(Proposal) to avoid having to handle `TemplateData` everywhere: https://github.com/kubermatic/kubermatic/pull/2145

File: api/pkg/controller/monitoring/monitoring_controller.go [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+)
+
+const (
+	// The monitoring controller waits for the cluster to become healthy,
+	// before adding the monitoring components to the clusters
+	healthCheckPeriod = 5 * time.Second
+)
+
+// UserClusterConnectionProvider offers functions to retrieve clients for the given user clusters
+type UserClusterConnectionProvider interface {
```
Comment (alvaroaleman) [inline api/pkg/controller/monitoring/monitoring_controller.go]: Why not use the existing `UserClusterConnectionProvider` type from the cluster controller?

File: api/cmd/kubermatic-controller-manager/controllers.go [hunk 7]
```diff
@@ -158,6 +160,46 @@ func createBackupController(ctrlCtx *controllerContext) (runner, error) {
...
+	}
+
+	return monitoring.New(
+		ctrlCtx.kubeClient,
+		client.New(ctrlCtx.kubeInformerFactory.Core().V1().Secrets().Lister()),
+
+		ctrlCtx.runOptions.dc,
+		dcs,
+		ctrlCtx.runOptions.overwriteRegistry,
+		ctrlCtx.runOptions.nodePortRange,
```
Comment (alvaroaleman) [inline api/cmd/kubermatic-controller-manager/controllers.go:181]: Cant we just leave this empty everywhere in the monitoring conroller as its not needed?

File: api/pkg/controller/monitoring/resources.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+func GetConfigMapCreators() []resources.ConfigMapCreator {
+	return []resources.ConfigMapCreator{
+		prometheus.ConfigMap,
+	}
+}
+
+func (c *Controller) ensureConfigMaps(cluster *kubermaticv1.Cluster) error {
+	creators := GetConfigMapCreators()
+
+	data, err := c.getClusterTemplateData(cluster)
```
Comment (alvaroaleman) [inline api/pkg/controller/monitoring/resources.go]: Wouldn't it be easier to make the `getClusterTemplateData` call inside the main sync loop and passing it into the `ensure...` funcs instead of doing it inside every single one of them?

File: api/pkg/controller/monitoring/resources_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+				objects = append(objects, test.cfgm)
+			}
+			controller := newTestController(objects, []runtime.Object{test.cluster})
+			beforeActionCount := len(controller.kubeClient.(*fake.Clientset).Actions())
+
+			if err := controller.ensureConfigMaps(test.cluster); err != nil {
+				t.Errorf("failed to ensure ConfigMap: %v", err)
+			}
+			if test.cfgm != nil {
+				if len(controller.kubeClient.(*fake.Clientset).Actions()) != beforeActionCount+1 && controller.kubeClient.(*fake.Clientset).Actions()[beforeActionCount].GetVerb() != "update" {
```
Comment (alvaroaleman) [inline api/pkg/controller/monitoring/resources_test.go]: Wouldn't it make more sense to put the expected actions on the testcase?

File: api/pkg/controller/monitoring/userclusterresources.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+	if len(c.Spec.MachineNetworks) > 0 {
+		creators = append(creators, ipamcontroller.ClusterRole)
+	}
+
+	return creators
+}
+
+func (cc *Controller) userClusterEnsureClusterRoles(c *kubermaticv1.Cluster) error {
+	client, err := cc.userClusterConnProvider.GetClient(c)
```
Comment (alvaroaleman) [inline api/pkg/controller/monitoring/userclusterresources.go]: Same as for the templateData, wouldn't it make sense to get the user cluser client in the main sync loop and pass it into the `userCluserEnsure....` instead of getting it in every single one of them?

**Review by alvaroaleman — CHANGES_REQUESTED** [review]

File: api/pkg/controller/monitoring/monitoring_controller.go (hunk 6 above)
Comment (mrIncompetent) [reply api/pkg/controller/monitoring/monitoring_controller.go]: I would live to avoid to import here from the cluster controller. Interfaces should (best case) be defined where they are used

File: api/cmd/kubermatic-controller-manager/controllers.go (hunk 7 above)
Comment (mrIncompetent) [reply api/cmd/kubermatic-controller-manager/controllers.go:181]: I would keep the fow now, until we have https://github.com/kubermatic/kubermatic/pull/2145 in

File: api/pkg/controller/monitoring/monitoring_controller.go (hunk 6 above)
Comment (alvaroaleman) [reply api/pkg/controller/monitoring/monitoring_controller.go]: Then they shouldn't be exported

**Review by alvaroaleman — DISMISSED** [review]

**Review by mrIncompetent — APPROVED** [review]

**Review by alvaroaleman — APPROVED** [review]
